### PR TITLE
[Feat] Implement area weights for grid-point averaging

### PR DESCRIPTION
**Pull request by prajwal-tech07**
Implemented area weights for grid-point averaging in the metrics module to ensure spatially weighted averaging instead of uniform averaging. Updated the related metric computations accordingly and added corresponding unit tests to validate correctness.

This improves the physical correctness of spatial averaging, especially for latitude-longitude grids where grid cells represent unequal surface areas.

No additional external dependencies were introduced.

## Issue Link
#212 – mask_and_reduce_metric does not support area-weighted averaging over grid points

## Type of change

- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📖 Documentation

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my feature works
- [x] I have given the PR a clear name

### CODE REVIEW

**Review by prajwal-tech07 — COMMENTED** [review]
reviewed

**Comment by sadamov** [comment]
This supercedes #256. @prajwal-tech07 please use the original PR template.

**Comment by Joltsy10** [comment]
Reviewed the test suite in #258. While the coverage is thorough, there's some overlap worth noting:

test_uniform_weights_equals_no_weights and test_no_weights_gives_same_result_as_before are essentially testing the same invariant. Since both rely on mask_and_reduce_metric, we could test the weighting logic once in TestMaskAndReduceMetric and use a single parametrized check at the metric level to verify the API hook exists.

TestComputeAreaWeights could also be condensed. Testing shape, dtype, positivity, and normalization in four separate tests adds boilerplate; I think one test with multiple assertions covers it just as well.

The tests that cover genuinely distinct behavior are test_metrics_with_batch_and_time_dims, test_weighted_mean_with_mask, and the manual calculation tests. A suite focused on those would be leaner and easier to maintain.

File: neural_lam/metrics.py [hunk 1]
```diff
@@ -18,17 +20,65 @@ def get_metric(metric_name):
     return DEFINED_METRICS[metric_name_lower]
 
 
-def mask_and_reduce_metric(metric_entry_vals, mask, average_grid, sum_vars):
+def compute_area_weights(datastore):
+    """
+    Compute area weights for each grid point based on the cosine of the
+    latitude. This accounts for the fact that grid cells at higher latitudes
+    cover less surface area than cells near the equator.
```
Comment (joeloskarsson) [inline neural_lam/metrics.py:27]: This is true for regular lat-lon grids, but not completely in general. Neural-lam supports all kinds of irregular grids where this might not be the case. While this is great functionality to have, and will be useful and necessary in most global setups, I would really like to make the use of such weights optional, to not cause issues in setups with other grids.

Comment (joeloskarsson) [reply neural_lam/metrics.py:27]: Perhaps the use of this (on/off) could be configures in the same config files that we now use to set the weight for each variable in the loss?

Comment (Joltsy10) [reply neural_lam/metrics.py:27]: Joel's point makes sense, keeping metrics.py generic with grid_weights as optional is the right direction. One question on where the weight computation itself should live. Should it stay in metrics, computed on the fly when needed, or should it be computed once at ARModel init and registered as a buffer alongside boundary_mask, with the config just controlling whether it's enabled? The buffer approach feels cleaner for consistency with how boundary_mask is handled, but wanted to check if you had a preference before Prajwal iterates on this.

Comment (santhil-cyber) [reply neural_lam/metrics.py:27]: Good question @Joltsy10 — I'd strongly lean toward the buffer approach (compute once at `ARModel.__init__`, register alongside `boundary_mask`).

A few reasons:

1. **Consistency** — `boundary_mask` is already a registered buffer. Treating `grid_weights` the same way keeps the pattern predictable.
2. **Performance** — `compute_area_weights()` involves a `cos()` call over all grid nodes. Small cost, but no reason to pay it every forward pass when weights are static.
3. **Config-driven** (addressing Joel's point) — if weights are a buffer, the config just controls whether to populate it or leave it `None`. The metrics code stays generic: `grid_weights=None` → unweighted, `grid_weights=tensor` → weighted.

On the edge cases — in my PR #316 I added two guards that might be worth incorporating here:

- **Device/dtype cast:** `grid_weights.to(metric_entry_vals.device, dtype=metric_entry_vals.dtype)` before any operations — prevents GPU/CPU mismatch during eval
- **Zero-sum guard:** `if w_sum <= 0: raise ValueError(...)` after masking — prevents silent `inf` from divide-by-zero if all weighted nodes get masked out

Happy to help align the implementations if that's useful.

**Comment by prajwal-tech07** [comment]
Thanks for all the feedback! I've addressed everything in the latest push:

@joeloskarsson — Area weights are now fully optional via a `use_area_weights: bool` flag in `TrainingConfig` (defaults to False), so irregular grids aren't affected. The setting lives in the same config alongside `state_feature_weighting` and `output_clamping`.

@Joltsy10 — Went with the buffer approach: `compute_area_weights()` is called once at `ARModel.__init__`( when enabled, registered alongside `boundary_mask`. When disabled, `self.grid_weights = None` and metrics stay unweighted. Also consolidated the redundant tests — merged the two backward-compat tests into one and condensed `TestComputeAreaWeights` into a single test with multiple assertions.

@santhil-cyber — Added `grid_weights.to(device=..., dtype=...)` before operations to prevent GPU/CPU mismatch, and a `if w_sum <= 0: raise ValueError(...)` guard after masking to prevent silent inf.

File: neural_lam/metrics.py (hunk 1 above)
Comment (joeloskarsson) [reply neural_lam/metrics.py:27]: Yea I think the buffer approach does make sense

File: neural_lam/metrics.py [hunk 2]
```diff
@@ -18,17 +20,65 @@ def get_metric(metric_name):
...
+    """
+    # Get projected x, y coordinates for state grid points
+    xy = datastore.get_xy(category="state", stacked=True)  # (N, 2)
+
+    # Transform projected coordinates to geographic lat/lon
+    lonlat = ccrs.PlateCarree().transform_points(
+        src_crs=datastore.coords_projection,
+        x=xy[:, 0],
+        y=xy[:, 1],
+    )  # (N, 3) -> [lon, lat, z]
```
Comment (joeloskarsson) [inline neural_lam/metrics.py:53]: Note that #168 will implement a `get_lat_lon` method for the datastore. So when that is merged that can also be used here.

File: neural_lam/metrics.py [hunk 3]
```diff
@@ -39,12 +89,36 @@ def mask_and_reduce_metric(metric_entry_vals, mask, average_grid, sum_vars):
...
     if average_grid:  # Reduce grid first
-        metric_entry_vals = torch.mean(
-            metric_entry_vals, dim=-2
-        )  # (..., d_state)
+        if grid_weights is not None:
+            # Weighted mean over grid dimension
+            # Reshape weights for broadcasting: (N', 1)
+            w = grid_weights.to(
+                device=metric_entry_vals.device,
+                dtype=metric_entry_vals.dtype,
```
Comment (joeloskarsson) [inline neural_lam/metrics.py:103]: Ensuring correct device and dtype should be responsibility of whoever calls this function.

File: neural_lam/metrics.py [hunk 4]
```diff
@@ -39,12 +89,36 @@ def mask_and_reduce_metric(metric_entry_vals, mask, average_grid, sum_vars):
...
+            w_sum = w.sum(dim=-2, keepdim=True)
+            if w_sum.item() <= 0:
+                raise ValueError(
+                    "Sum of grid weights is non-positive after masking. "
+                    "All weighted grid points may have been masked out."
+                )
+            w = w / w_sum
+            metric_entry_vals = torch.sum(
+                metric_entry_vals * w, dim=-2
+            )  # (..., d_state)
```
Comment (joeloskarsson) [inline neural_lam/metrics.py:117]: I think a better approach here is to keep this as a mean, and not make the weights sum to 1. Instead when created we should rescale the weights with the average, so their mean is 1. See for example weatherbench 2 for this.

Comment (joeloskarsson) [reply neural_lam/metrics.py:117]: If this is not a mean (first weighted by given grid weights), then the `average_grid` argument doesn't make sense.

File: tests/test_metrics.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    # They should generally differ (we use a loose check — if they are
+    # identical bit-for-bit that would be extremely unlikely with random data)
+    # This is a soft assertion; we mainly verify correctness elsewhere.
+    assert result_weighted.shape == result_no_weights.shape
+
+
+# ============================================================
+# Test: mask_and_reduce_metric directly
+# ============================================================
+class TestMaskAndReduceMetric:
```
Comment (joeloskarsson) [inline tests/test_metrics.py:111]: Why structure this in a class, seems inconsistent with how we typically structure tests.

File: neural_lam/config.py [hunk 6]
```diff
@@ -103,6 +103,8 @@ class TrainingConfig:
         default_factory=OutputClamping
     )
 
+    use_area_weights: bool = False
```
Comment (joeloskarsson) [inline neural_lam/config.py:106]: I suggest a different name, as we here assume that we can compute this area from latitude.
```suggestion
    use_lat_area_weights: bool = False
```

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Starting to look pretty good! Great discussions here. I added some suggestions 😄

File: neural_lam/metrics.py [hunk 7]
```diff
@@ -39,12 +89,36 @@ def mask_and_reduce_metric(metric_entry_vals, mask, average_grid, sum_vars):
...
+        # Also mask the grid weights if provided
+        if grid_weights is not None:
+            grid_weights = grid_weights[mask]  # (N',)
 
     # Optionally reduce last two dimensions
     if average_grid:  # Reduce grid first
-        metric_entry_vals = torch.mean(
-            metric_entry_vals, dim=-2
-        )  # (..., d_state)
+        if grid_weights is not None:
```
Comment (Omnagar) [inline neural_lam/metrics.py:98]: Small design question regarding the weighting logic here.

Since `compute_area_weights()` already normalizes the weights so that
their mean equals 1, would it still be necessary to renormalize them
again inside `mask_and_reduce_metric()` after masking?

For example, if masking removes some grid points, the normalization here
changes the relative scaling of the remaining weights. I’m wondering if
it might be clearer to keep the weights fixed (mean≈1) and simply apply
a weighted mean without renormalizing after masking.

Curious what the preferred behaviour should be in this case.

**Review by Omnagar — COMMENTED** [review]
Thanks for the detailed implementation of area-weighted metrics.

I left a small comment in `metrics.py` regarding the normalization of grid weights after masking. I was curious whether renormalizing inside `mask_and_reduce_metric()` is necessary given that `compute_area_weights()` already normalizes the weights.

Overall the approach looks clean and the integration with the existing metric pipeline makes sense.

File: neural_lam/metrics.py (hunk 7 above)
Comment (Joltsy10) [reply neural_lam/metrics.py:98]: The renormalization after masking is intentional. compute_area_weights() normalizes over all grid nodes, but after the interior mask is applied the remaining weights no longer sum to 1 since boundary nodes have been removed. Without renormalizing, the weighted mean would not integrate correctly over the interior and the loss scale would vary depending on boundary region size. The second normalization ensures the result is a proper weighted average over interior nodes regardless of how large the boundary region is.

Comment (joeloskarsson) [reply neural_lam/metrics.py:98]: related to https://github.com/mllam/neural-lam/pull/258#discussion_r2937039832

Comment (Omnagar) [reply neural_lam/metrics.py:98]: > The renormalization after masking is intentional. compute_area_weights() normalizes over all grid nodes, but after the interior mask is applied the remaining weights no longer sum to 1 since boundary nodes have been removed. Without renormalizing, the weighted mean would not integrate correctly over the interior and the loss scale would vary depending on boundary region size. The second normalization ensures the result is a proper weighted average over interior nodes regardless of how large the boundary region is.

That makes sense, thanks for the clarification — especially the point about
the loss scale changing depending on the masked region size.

One follow-up thought: would it be worth documenting this behaviour explicitly
in the docstring of `mask_and_reduce_metric()`? It feels like a subtle detail,
and future contributors might assume the weights are only normalized once at
creation time.

I can help add a short clarification there if that sounds useful.

**Comment by sadamov** [comment]
Closing this in favour of #650, which consolidates the global forecasting / spherical domain track for v0.8.0. The cos-lat area-weighted grid-point averaging here will be folded into the consolidated PR. Thanks for the contribution @prajwal-tech07, you'll be credited.
